kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt

```diff
@@ -64,6 +69,7 @@ internal open class ArrayChannel<E>(
                         }
                     }
                 }
+                ensureCapacity(size)
                 buffer[(head + size) % capacity] = element // actually queue element
```

Comment:
Here it should use `buffer.size` as a modulo instead of `capacity` when storing to the buffer. Also, in all other places in the code `foo % capacity` shall be replaced with `foo % buffer.size` (all the code that reads/write to/from `buffer` and the code that updates `head`).

Reply:
Shame, fixed